### PR TITLE
Provide the proof-of-concept for CV skills extraction

### PR DESCRIPTION
**Pull request by lilsweetcaligula**
#37 

Please do not merge this PR until a detailed discussion has been taken place. This PR is meant to serve as a rough example.

The skills have been hard-coded for now but it really does not need to be this way.

The testing of this PR is blocked by issue #43.

### CODE REVIEW

File: src/candidates/core/services/cv-parsing.service.js [hunk 1]
```diff
@@ -0,0 +1,572 @@
+import similarity from 'string-similarity';
+
```
Comment (lilsweetcaligula) [inline src/candidates/core/services/cv-parsing.service.js]: Please skip to the definition of `fetchGoodMatchesIn`, for this is where the algorithm is implemented.

Note to self: these other utility functions could be extracted into the modules of their own.

**Comment by ivan-kolmychek-devv** [comment]
Thank you for the pull-request! :tada: 

This looks well thought-out and is quite well documented. I think it may benefit from some splitting into smaller files, but that's more of a question of taste. :slightly_smiling_face: 

Keeping in mind that it is a proof-of-concept, I have few questions:

_(disclaimer: please keep in mind that you probably have thought it through much better than I did - some of the questions can prove to be ignorant to some degree :slightly_smiling_face: )_

- Do we have any better way to distinguish where skills start/end? Are there any assumptions we can make to make it easier, like "Skills usually go after `<one of the appropriate headers>` and are separated by `<one of the appropriate separators>` each"?

- Without the optimization that you mention in comments, how does it behave performance-wise? In your opinion.

- In general, after this PoC, what do you think about current approach of `extract text -> extract stuff from text`? Would you suggest to still roll with it, or look for other options?

**Comment by lilsweetcaligula** [comment]
> Do we have any better way to distinguish where skills start/end? Are there any assumptions we can make to make it easier, like "Skills usually go after <one of the appropriate headers> and are separated by <one of the appropriate separators> each"?

Please see this answer in this discussion: https://github.com/DevvJobs/candidates-web/issues/37#issuecomment-476438953

> Without the optimization that you mention in comments, how does it behave performance-wise? In your opinion.

My guess would be around a 15% performance increase, but it's a guess. With the brute-force approach, assuming the text is tokenized based on a word, we would have to traverse the text from start to finish for each skill. The optimization drops some of the reads, by providing a given skill with a list of possible matches to consider.

> In general, after this PoC, what do you think about current approach of extract text -> extract stuff from text? Would you suggest to still roll with it, or look for other options?

I'd say, it is always a good idea to look for better options. But for the beta release of the feature, I think it need not be perfect as long as it's easy to swap out for the better alternatives, once they are found. This current solution is probably OK for the time-being. Is it good? I am not sure. But I do know that I am currently suffering from tunnel vision, so it's important to continue brainstorming together (just like we are now).

---

Regarding this algorithm, I'd probably redo it, because it currently it yields quite a few false positives and quite a few false negatives. But that could be due to https://github.com/DevvJobs/candidates-web/issues/45

I was thinking, what if the file with the list of skills looks like this:
```
[
  {
    skill: 'React',
    aliases: ['React.js']
  },
  {
    skill: 'JavaScript',
    aliases: ['JS', 'ES', 'EcmaScript']
  },
  ...
]
```

Then tokenize the content of the CV, don't bother with optimizations at this point, go through each skill, and check if the text:
- has a set of subsequent words that fuzzy-include the skill
- has a set of subsequent words that fuzzy-includes any of its aliases

Once all of the skills are collected, we try to fetch them from backend.

Then we should probably also limit the size of the CV, too.

**Comment by lilsweetcaligula** [comment]
This is a bit off-topic, but once GitHub integration is implemented, you could alternatively let users choose to fetch their skills based on their GitHub account.

**Comment by ivan-kolmychek-devv** [comment]
> Please see this answer in this discussion: #37 (comment)

:+1: Replied there.

> My guess would be around a 15% performance increase, but it's a guess. With the brute-force approach, assuming the text is tokenized based on a word, we would have to traverse the text from start to finish for each skill. The optimization drops some of the reads, by providing a given skill with a list of possible matches to consider.

Sorry, I think I asked it in an unclear way. What I meant to ask is - according to your testing, do you think it behaves in a good enough way to be useful, when it comes to performance? Like in "does not eat all the ram", "does not hog CPU forever" and so on.

To clarify, neither me nor @tetiana-kolmychek-devv noticed any problems with performance, but we also assume that you tested it more than us and on various amounts of skills.

> Regarding this algorithm, ...

If you think that is the better way to do so - go for it. :slightly_smiling_face:

**Comment by ivan-kolmychek-devv** [comment]
Me and @tetiana-kolmychek-devv  checked out into the branch and tried it out on few CVs that we had, both self-made and exported from other systems.

Our notes:

- A lot of text is missing whitespaces in parsing result, as you already reported in #45. It leads to a lot of false-negatives.

- Final output with `console.log` just display `Array(n)` in the console. We needed to add `.join(',')` to see the actual result.

- It may be a good idea to also output the intermediary results, for example - after removal of punctuation marks.

- If you put two or more skills with `/`, they are not getting parsed. Examples: `HTML/CSS`, `Elixir/Phoenix`, `Ruby/Rails`.

- A lot of words triggered false positives - we got `devops`, `junit` and `unity` false-detected in two of the CVs. We do not know which exact words triggered it, we assume it was `developer`, `jquery` and something else.

- With `yarn candidates`, we also got the #43 (that you have already reported :tada:) and it broke the import for us completely. Had to use `yarn build` + `yarn static` to actually test it out.

Looks good as a start, but we think it's not accurate enough yet, at least not for us. Hope it can get better. :+1:

**Comment by lilsweetcaligula** [comment]
Could you please assign a `help-wanted` label to all issues involving CV parsing? Thank you!

**Comment by ivan-kolmychek-devv** [comment]
@lilsweetcaligula we don't have `help-wanted` label because almost everything is "help wanted", except from few issues. :slightly_smiling_face: 

If you cannot continue working on them for any reason, that is OK. Please confirm so we can consider them free for taking. 

Again, thank you for the already made contribution. :+1: :rocket:

**Comment by lilsweetcaligula** [comment]
@ivan-kolmychek-devv Yep, confirmed. Free for taking. I will probably do some of the work on the issue when I have the time, but I cannot make... promises. :trollface:

**Comment by ivan-kolmychek-devv** [comment]
@lilsweetcaligula thank you for the confirmation. :+1:  Good luck and have fun, we'll be glad to see you again. :wave: 

I'm closing this PR but it can be used as base or reference material for anyone else who may want to pick up the issue. :octocat:

**Pull request closed without merging**